client/components/voip/User.ts

```diff
@@ -0,0 +1,424 @@
+/**
+ * Class representing SIP UserAgent
+ * @remarks
+ * This class encapsulates all the details of sip.js and exposes
+ * a very simple functions and callback handlers to the outside world.
+ * This class thus abstracts user from Browser specific media details as well as
+ * SIP specific protol details.
+ */
+
+import {
+	UserAgent,
+	UserAgentOptions,
+	UserAgentDelegate,
+	Invitation,
+	InvitationAcceptOptions,
+	Session,
+	SessionState,
+	Registerer,
+} from 'sip.js';
+import { OutgoingRequestDelegate } from 'sip.js/lib/core';
+import { SessionDescriptionHandler } from 'sip.js/lib/platform/web';
+
+import { ICallEventDelegate } from './CallEventDelegate';
+import { CallState } from './Callstate';
+import { IConnectionDelegate } from './ConnectionDelegate';
+import { Operation } from './Operations';
+import { IRegisterHandlerDeligate } from './RegisterHandlerDelegate';
+import Stream from './media/Stream';
+import { Logger } from './utils/Logger';
+// User state is based on whether the User has sent an invite(UAC) or it
+// has received an invite (UAS)
+enum UserState {
+	IDLE,
+	UAC,
+	UAS,
+}
+export class User implements UserAgentDelegate, OutgoingRequestDelegate {
+	private connectionDelegate: IConnectionDelegate;
+
+	private registrationDelegate: IRegisterHandlerDeligate;
+
+	private callEventDelegate: ICallEventDelegate;
+
+	private session: Session | undefined;
+
+	private remoteStream: Stream | undefined;
+
+	private config: any = {};
```

Comment:
I thought about it. Contemplated and then stopped. Because interface would not allow to have snake_case. And I had affinity to pass snake case for configuration. This was to distinguish our configuration parameters from sip.js's configuration parameter. Later changed my mind. Now we have an interface for VoIPUser's configuration parameters.